lib/eventsource/eventsource.js

```diff
@@ -0,0 +1,326 @@
+   * values described below.
+   * @returns {0|1|2}
+   * @readonly
+   */
+  get readyState () {
+    return this.#readyState
+  }
+
+  /**
+   * Returns the URL providing the event stream.
+   * @readonly
+   * @returns {string}
+   */
+  get url () {
+    return this.#url
+  }
+
+  /**
+   * Returns a boolean indicating whether the EventSource object was
+   * instantiated with CORS credentials set (true), or not (false, the default).
+   */
+  get withCredentials () {
+    return this.#withCredentials
+  }
+
+  async #connect () {
+    this.#readyState = CONNECTING
+    this.#connection = null
+
+    /**
+     * @type {RequestInit}
+     */
+    const options = {
+      method: 'GET',
+      redirect: 'manual',
+      keepalive: true,
+      headers: {
+        Accept: mimeType,
+        'Cache-Control': 'no-cache',
+        Connection: 'keep-alive'
+      },
+      signal: this.#controller.signal
+    }
+
+    if (this.#lastEventId) {
+      options.headers['Last-Event-ID'] = this.#lastEventId
+    }
+
+    options.credentials = this.#withCredentials ? 'include' : 'omit'
+
+    try {
+      this.#connection = await fetch(this.#url, options)
+
+      // Handle HTTP redirects
+      // https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events-intro
+      switch (this.#connection.status) {
+        // Redirecting status codes
+        case 301: // 301 Moved Permanently
+        case 302: // 302 Found
+        case 307: // 307 Temporary Redirect
+        case 308: // 308 Permanent Redirect
+          if (!this.#connection.headers.has('Location')) {
+            this.close()
+            this.dispatchEvent(new Event('error'))
+            return
+          }
+          this.#url = new URL(this.#connection.headers.get('Location'), new URL(this.#url).origin).href
+          this[kState].origin = new URL(this.#url).origin
+          this.#connect()
+          return
+        case 204: // 204 No Content
+          // Clients will reconnect if the connection is closed; a client can be told to stop reconnecting
+          // using the HTTP 204 No Content response code.
+          this.close()
+          this.dispatchEvent(new Event('error'))
+          return
+        case 200:
+          if (this.#connection.headers.get('Content-Type') !== mimeType) {
+            this.close()
+            this.dispatchEvent(new Event('error'))
+            return
+          }
+          break
+        default:
+          this.close()
+          this.dispatchEvent(new Event('error'))
+          return
+      }
+
+      if (this.#connection === null) {
+        this.close()
+        this.dispatchEvent(new Event('error'))
+        return
+      }
+
+      const self = this
+
+      pipeline(this.#connection.body,
+        new EventSourceStream({
+          eventSourceState: this[kState],
+          push: function push (chunk) {
+            self.dispatchEvent(chunk)
+          }
```

Comment:
```suggestion
          push: (chunk) => {
            this.dispatchEvent(chunk)
          }
```

can get rid of self afterwards